File: takengine/mapengine/sdk/src/raster/mobac/MobacMapSourceTilesetSupport.cpp

#include <algorithm>

#include "raster/mobac/MobacMapSourceTilesetSupport.h"

#include "raster/DatasetDescriptor.h"
#include "raster/ImageDatasetDescriptor.h"
#include "raster/mobac/EquirectangularMobacMapSourceTilesetSupport.h"
#include "raster/mobac/MobacMapSource.h"
#include "raster/mobac/MobacMapSourceFactory.h"
#include "raster/mobac/MobacTileClient.h"
#include "raster/mobac/WebMercatorMobacMapSourceTilesetSupport.h"
#include "raster/tilepyramid/TilesetInfo.h"
#include "db/Database.h"
#include "db/Statement.h"
#include "util/FutureTask.h"
#include "util/IO2.h"

using namespace atakmap::raster::mobac;

using namespace atakmap::db;
using namespace atakmap::raster;
using namespace atakmap::raster::tilepyramid;
using namespace atakmap::renderer;
using namespace atakmap::util;

/*
 protected final MobacMapSource mapSource;
 protected final TilesetInfo tsInfo;
 protected long defaultExpiration;
 protected boolean checkConnectivity;
 protected MobacTileClient client;
 */

atakmap::raster::tilepyramid::TilesetSupport::Spi *const MobacMapSourceTilesetSupport::SPI = createSpi();


namespace
{
    class CachingBitmapLoader : public FutureTaskImpl<Bitmap> {
    public:
        CachingBitmapLoader(MobacTileClient *client, MobacMapSource *source, int latIndex, int lngIndex, int level, bool checkConnectivity/*, BitmapFactory.Options opts*/);
        virtual ~CachingBitmapLoader();
    protected:
        virtual bool runImpl();
    
    private:
        MobacTileClient * const client;
        MobacMapSource * const source;
        const int latIndex;
        const int lngIndex;
        const int level;
        const bool checkConnectivity;
        /*private final BitmapFactory.Options opts;*/
    };
    
    class SpiImpl : public TilesetSupport::Spi {
    public:
        SpiImpl();
    public:
        virtual const char *getName() const;
        virtual TilesetSupport *create(TilesetInfo *tsInfo, AsyncBitmapLoader *bitmapLoader);
    };
}

MobacMapSourceTilesetSupport::MobacMapSourceTilesetSupport(TilesetInfo *info, MobacMapSource *source)
: tsInfo(info),
mapSource(source),
defaultExpiration(strtoll(DatasetDescriptor::getExtraData(*info->getInfo(), "defaultExpiration", "-1"), NULL, 10)),
checkConnectivity(true),
client(nullptr)
{}

MobacMapSourceTilesetSupport::~MobacMapSourceTilesetSupport() { }

void MobacMapSourceTilesetSupport::setOfflineMode(bool offlineOnly)
{
    this->client->setOfflineMode(offlineOnly);
}

bool MobacMapSourceTilesetSupport::isOfflineMode()
{
    return this->client->isOfflineMode();
}

// XXX-- general helpers stuff. Do we have something better, or refactor to share?
namespace  {
    struct FindStringPred {
        FindStringPred(const char *staticStr)
        : str(staticStr) { }
        inline bool operator()(const PGSC::String &item) const {
            return strcmp(item, str) == 0;
        }
        const char *str;
    };

    static bool vectorContains(const std::vector<PGSC::String> &vec, const char *str) {
        return std::find_if(vec.begin(), vec.end(), FindStringPred(str)) != vec.end();
    }
}

void MobacMapSourceTilesetSupport::init()
{
    const char *offlineCache = this->tsInfo->getInfo()->getExtraData("offlineCache");
    TAK::Engine::Port::String pathStr;
    
    //XXX-- emergency fix for iOS cache database open failure
#if 0
    if (offlineCache != nullptr) {
        pathStr = offlineCache;
        TAK::Engine::Util::File_getRuntimePath(pathStr);
        
        if (!atakmap::util::isFile(pathStr.get())) {
            atakmap::util::Logger::log(atakmap::util::Logger::Error, "MobacMapSourceTilesetSupport: offlineCache database does not exist");
        }
        
        std::auto_ptr<Database> database(db::openDatabase(pathStr.get()));
        std::vector<PGSC::String> tables = db::getTableNames(*database);
        
        if (!vectorContains(tables, "tiles")) {
            database->execute("CREATE TABLE tiles (key INTEGER PRIMARY KEY, provider TEXT, tile BLOB)");
        }
        if (!vectorContains(tables, "ATAK_catalog")) {
            database->execute("CREATE TABLE ATAK_catalog (key INTEGER PRIMARY KEY, access INTEGER, expiration INTEGER, size INTEGER)");
        }
        if (!vectorContains(tables, "ATAK_metadata")) {
            database->execute("CREATE TABLE ATAK_metadata (key TEXT, value TEXT)");
            std::auto_ptr<Statement> stmt(database->compileStatement("INSERT INTO ATAK_metadata (key, value) VALUES('srid', ?)"));
            stmt->bind(1, this->tsInfo->getInfo()->getSpatialReferenceID());
            stmt->execute();
        }
    }
#endif
    
    this->mapSource->clearAuthFailed();
    this->client = std::unique_ptr<MobacTileClient>(new MobacTileClient(this->mapSource.get(), pathStr.get()));
}

void MobacMapSourceTilesetSupport::release()
{
    if (this->client != nullptr)
        this->client->close();
}

void MobacMapSourceTilesetSupport::start()
{
    this->checkConnectivity = true;
}

void MobacMapSourceTilesetSupport::stop()
{
}

FutureTask<Bitmap> MobacMapSourceTilesetSupport::getTile(int latIndex, int lngIndex, int level/*, BitmapFactory.Options opts*/)
{
    FutureTask<Bitmap> r(PGSC::RefCountablePtr<FutureTaskImpl<Bitmap>>(new CachingBitmapLoader(this->client.get(),
                                                                                               this->mapSource.get(),
                                                                                               latIndex, lngIndex,
                                                                                               level,
                                                                                               this->checkConnectivity/*, opts*/)));
    this->checkConnectivity = false;
    return r;
}

TilesetSupport::Spi *MobacMapSourceTilesetSupport::createSpi()
{
    return new SpiImpl();
}

namespace
{
    CachingBitmapLoader::CachingBitmapLoader(MobacTileClient *c, MobacMapSource *src, int lat, int lng, int lvl, bool checkConn/*, BitmapFactory.Options opts*/) :
    client(c),
    source(src),
    latIndex(lat),
    lngIndex(lng),
    level(lvl),
    checkConnectivity(checkConn)
    {}
    
    CachingBitmapLoader::~CachingBitmapLoader() {
        
    }
    
    bool CachingBitmapLoader::runImpl() {
        
        if (this->checkConnectivity)
            this->source->checkConnectivity();
        
        Bitmap bitmap;
        if (this->client->loadTile(&bitmap, this->level, this->lngIndex, this->latIndex/*, this.opts*/, nullptr)) {
            return this->completeProcessing(bitmap);
        } else {
            this->setState(SharedState::Error);
        }
        
        return false;
    }
    
    SpiImpl::SpiImpl() {}
    
    const char *SpiImpl::getName() const {
        return "mobac";
    }
    
    TilesetSupport *SpiImpl::create(TilesetInfo *info, AsyncBitmapLoader *loader) {
        atakmap::util::Logger::log(atakmap::util::Logger::Debug, "creating Mobac TilesetSupport for %s", info->getInfo()->getURI());
        MobacMapSource *mapSource = MobacMapSourceFactory::create(info->getInfo()->getURI());
        
        if (mapSource == nullptr)
            return nullptr;
        
        std::unique_ptr<MobacMapSourceTilesetSupport> impl(new MobacMapSourceTilesetSupport(info, mapSource));
        if (info->getInfo()->getSpatialReferenceID() == 4326)
            return new EquirectangularMobacMapSourceTilesetSupport(info, loader, impl.release());
        else if (info->getInfo()->getSpatialReferenceID() == 3857)
            return new WebMercatorMobacMapSourceTilesetSupport(info, loader, impl.release());
        else
            return nullptr;
    }
}

File: takengine/takengine/src/main/java/com/atakmap/map/layer/raster/mobac/MobacMapSourceTilesetSupport.java

package com.atakmap.map.layer.raster.mobac;

import java.io.File;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.FutureTask;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.atakmap.android.maps.graphics.GLBitmapLoader;
import com.atakmap.coremap.io.IOProviderFactory;
import com.atakmap.database.DatabaseIface;
import com.atakmap.database.Databases;
import com.atakmap.android.maps.tilesets.OnlineTilesetSupport;
import com.atakmap.android.maps.tilesets.TilesetInfo;
import com.atakmap.android.maps.tilesets.TilesetSupport;
import com.atakmap.database.StatementIface;
import com.atakmap.map.projection.EquirectangularMapProjection;
import com.atakmap.map.projection.WebMercatorProjection;
import com.atakmap.coremap.filesystem.FileSystemUtils;
import com.atakmap.coremap.log.Log;

public final class MobacMapSourceTilesetSupport implements OnlineTilesetSupport {

    protected final MobacMapSource mapSource;
    protected final TilesetInfo tsInfo;
    protected boolean checkConnectivity;
    protected MobacTileClient client;

    private boolean offlineMode;
    private int tilesVersion = 0;

    public static final String TAG = "MobacMapSourceTilesetSupport";

    MobacMapSourceTilesetSupport(TilesetInfo tsInfo, MobacMapSource mapSource) {
        this.tsInfo = tsInfo;
        this.mapSource = mapSource;

        this.checkConnectivity = true;

        this.client = null;
        
        this.offlineMode = false;
    }

    /**************************************************************************/
    // Online Tileset Support

    @Override
    public final synchronized void setOfflineMode(boolean offlineOnly) {
        if(this.client != null)
            this.client.setOfflineMode(offlineOnly);
        if(offlineOnly != this.offlineMode)
            this.tilesVersion++;
        this.offlineMode = offlineOnly;
    }

    @Override
    public final synchronized boolean isOfflineMode() {
        return this.offlineMode;
    }

    /**************************************************************************/
    // Tileset Support

    public final void init() {

        String offlineCache = this.tsInfo.getInfo().getExtraData("offlineCache");

        if (offlineCache != null) {
            DatabaseIface database = null;
            try {
                database = IOProviderFactory.createDatabase(new File(offlineCache));
                Set<String> tables = Databases.getTableNames(database);
                if (!tables.contains("tiles")) {
                    database.execute("CREATE TABLE tiles (key INTEGER PRIMARY KEY, provider TEXT, tile BLOB)", null);
                }
                if (!tables.contains("ATAK_catalog")) {
                    database.execute("CREATE TABLE ATAK_catalog (key INTEGER PRIMARY KEY, access INTEGER, expiration INTEGER, size INTEGER)", null);
                }
                if (!tables.contains("ATAK_metadata")) {
                    database.execute("CREATE TABLE ATAK_metadata (key TEXT, value TEXT)", null);
                    StatementIface stmt = null;
                    try {
                        stmt = database.compileStatement("INSERT INTO ATAK_metadata (key, value) VALUES(?, ?)");
                        stmt.bind(1, "srid");
                        stmt.bind(2, String.valueOf(this.tsInfo.getInfo().getSpatialReferenceID()));
                        stmt.execute();
                    } finally {
                        if(stmt != null)
                            stmt.close();
                    }
                }
            } catch (Exception e) {
                // received a could not open database exception and a hard crash, try to recover.
                Log.d(TAG, "exception occurred opening/using: " + offlineCache, e);
                try { 
                    FileSystemUtils.delete(offlineCache);
                    // todo, cycle back to the initialization code.   
                } catch (Exception ioe) { 
                    Log.d(TAG, "error occurred deleting cache: " + offlineCache, ioe);
                } 
                offlineCache = null;
            } finally {
                if (database != null)
                    database.close();
            }
        }

        this.mapSource.clearAuthFailed();
        synchronized(this) {
            this.client = new MobacTileClient(this.mapSource, offlineCache);
            this.client.setOfflineMode(this.offlineMode);
        }
    }

    public final synchronized void release() {
        if (this.client != null)
            this.client.close();
    }

    public final void start() {
        this.checkConnectivity = true;
    }

    public final void stop() {
    }

    public final FutureTask<Bitmap> getTile(int latIndex, int lngIndex,
            int level, BitmapFactory.Options opts) {

        FutureTask<Bitmap> r = new FutureTask<Bitmap>(new CachingBitmapLoader(latIndex,
                lngIndex, level, this.checkConnectivity, opts));
        this.checkConnectivity = false;
        return r;
    }

    /**************************************************************************/

    public static class Spi implements TilesetSupport.Spi {
        public final static TilesetSupport.Spi INSTANCE = new Spi();

        @Override
        public String getName() {
            return "mobac";
        }

        @Override
        public TilesetSupport create(TilesetInfo info, GLBitmapLoader loader) {
            MobacMapSource mapSource = null;
            try {
                mapSource = MobacMapSourceFactory.create(new File(info.getInfo().getUri()));
            } catch (Exception e) {
                Log.d(TAG, "exception occurred: ", e);
            }

            if (mapSource == null)
                return null;

            MobacMapSourceTilesetSupport impl = new MobacMapSourceTilesetSupport(info, mapSource);
            if (info.getInfo().getSpatialReferenceID() == EquirectangularMapProjection.INSTANCE
                    .getSpatialReferenceID())
                return new EquirectangularMobacMapSourceTilesetSupport(info, loader, impl);
            else if (info.getInfo().getSpatialReferenceID() == WebMercatorProjection.INSTANCE
                    .getSpatialReferenceID())
                return new WebMercatorMobacMapSourceTilesetSupport(info, loader, impl);
            else
                return null;
        }
    }

    public int getTilesVersion(int latIndex, int lngIndex, int level) {
        return this.tilesVersion;
    }
    
    /**************************************************************************/

    private class CachingBitmapLoader implements Callable<Bitmap> {
        private final int latIndex;
        private final int lngIndex;
        private final int level;
        private final boolean checkConnectivity;
        private final BitmapFactory.Options opts;

        public CachingBitmapLoader(int latIndex, int lngIndex, int level,
                boolean checkConnectivity, BitmapFactory.Options opts) {
            this.latIndex = latIndex;
            this.lngIndex = lngIndex;
            this.level = level;
            this.checkConnectivity = checkConnectivity;
            this.opts = opts;
        }

        @Override
        public Bitmap call() throws Exception {
            if (this.checkConnectivity)
                MobacMapSourceTilesetSupport.this.mapSource.checkConnectivity();

            return MobacMapSourceTilesetSupport.this.client.loadTile(
                    this.level, this.lngIndex, this.latIndex, this.opts, null);
        }
    }
}
